haystack/schema.py

```diff
@@ -94,17 +96,52 @@ def __str__(self):
         return str(self.to_dict())
 
 
+@dataclass
```

Comment:
Have we decided about this 

> Decide if to go for plain class, dataclass, BaseModel or TypedDict

There is inconsistency across package about use of these class. Better to decide now and use it uniformly across code base.

Reply:
BTW FastAPI support dataclass but it use pydantic underneath https://fastapi.tiangolo.com/sq/advanced/dataclasses/

Reply:
We haven't made the decision yet for which style to go. Dataclasses are a good candidate - however some of our classes need custom inits / to_dict methods etc., so the value of Dataclasses bringing this out-of-the-box is not really big in these cases.
 
Yep, we know about the FAST API support (already using it in this early draft for `Answer` in the API) and are in contact with @tiangolo :)

Reply:
We had an internal discussion: on the short list are dataclasses and pydantic base models. 
Pro basemodel: runtime type checks, 
Con: we experienced significant overhead in the past (10-15% slower inference in FARM)

Reply:
Curious about reason for this -

> we experienced significant overhead in the past (10-15% slower inference in FARM)

Reply:
This is the related FARM PR: https://github.com/deepset-ai/FARM/pull/410

We didn't have time to investigate it properly back then, but if I remember correctly the instantiation of objects was slower and as one inference call can easily involve the init of hundreds of objects under the hood it became significant overhead

Reply:
Did a quick test and found pydantic again to be significantly slower. I also found a section related in the [docs](https://pydantic-docs.helpmanual.io/usage/models/#creating-models-without-validation) pointing towards it: 
> ... construct() is generally around 30x faster than creating a model with full validation

My mini test run showed ~ 3x for dataclass. In theory, you can disable the validation via the usage of construct(). However, then we don't gain much from using pydantic IMO.

So I would continue with dataclass here. Any objections @oryx1729 ?

```
  from haystack.schema import DocumentPydantic
    import time
    dicts = dicts * 100
    dicts = dicts[:100_000]

    for i in range(5):
        tic = time.time()
        documents = [DocumentPydantic.from_dict(d) for d in dicts]
        toc = time.time()
        print(f"Docs: {len(documents)} Time: {toc-tic}")

    # document_store.write_documents(dicts)
    
    # dataclass
    Docs: 100000
    Time: 0.4848208427429199
    Docs: 100000
    Time: 0.5104725360870361
    Docs: 100000
    Time: 0.4244520664215088
    Docs: 100000
    Time: 0.4213542938232422
    Docs: 100000
    Time: 0.4138145446777344

    # Pydantic
    Docs: 100000
    Time: 1.5547969341278076
    Docs: 100000
    Time: 1.5462539196014404
    Docs: 100000
    Time: 1.3664710521697998
    Docs: 100000
    Time: 1.4674921035766602
    Docs: 100000
    Time: 1.4913456439971924 
```